parachain/primitives/xcm-support/src/lib.rs

```diff
@@ -85,10 +88,18 @@ where
 		let weight = T::Weigher::weight(&mut message)
 			.map_err(|_| DispatchError::Other("Unweighable message."))?;
 
-		T::XcmExecutor::execute_xcm_in_credit(origin_location, message, weight, weight)
-			.ensure_complete()
-			.map_err(|_| DispatchError::Other("Xcm execution failed."))?;
+		with_transaction(|| {
+			let result =
+				T::XcmExecutor::execute_xcm_in_credit(origin_location, message, weight, weight)
+					.ensure_complete();
 
-		Ok(())
+			match result {
+				Ok(()) => TransactionOutcome::Commit(Ok(())),
+				Err(_) =>
+					return TransactionOutcome::Rollback(Err(DispatchError::Other(
+						"Xcm execution failed.",
```

Comment:
If `result` is an error we should be passing it to `TransactionOutcome::Rollback` so that it can handled or logged higher up the stack.

Reply:
The issue with this is that `DispatchResult` does not provide a way to accept an arbitrary Error type or an error message string that is built dynamically at runtime. It only accepts `&'static str`.

I have considered adding an enum type to represent errors returned from `transfer_reserve`.
```rust
pub enum TransferError {
    ZeroFees,
    UnweighableMessage,
    XcmExecution(xcm::v2::traits::Error),
}
```

The error here is of type `xcm::v2::traits::Error` and to pass this higher up the stack we would need to add it to the public interface of `reserve_transfer`. This would add a dependency on the `xcm` crate for `primitives/core` crate as well as `eth-app` and `erc20-app` pallets. Currently only `xcm-support` crate depends on the `xcm` crate.

Should we be concerned about adding these dependencies? We could potentially just log the error here and return `Err(DispatchError::Other("Xcm execution failed.")` as per the existing implementation.

Reply:
Cool, understood. Anyway, with the other changes you made (like hoisting with_transaction), I guess my earlier comment is obsolete.